icu4c/source/common/unicode/utfiter.h

```diff
@@ -0,0 +1,2045 @@
+                    (c < 0xf0 ?  // U+0800..U+FFFF except surrogates
+                        U8_LEAD3_T1_BITS[c &= 0xf] & (1 << ((t = *p) >> 5)) &&
+                        (t &= 0x3f, 1)
+                    :  // U+10000..U+10FFFF
+                        (c -= 0xf0) <= 4 &&
+                        U8_LEAD4_T1_BITS[(t = *p) >> 4] & (1 << c) &&
+                        (c = (c << 6) | (t & 0x3f), ++length, ++p != limit) &&
+                        (t = *p - 0x80) <= 0x3f) &&
+                    // valid second-to-last trail byte
+                    (c = (c << 6) | t, ++length, ++p != limit)
+                :  // U+0080..U+07FF
+                    c >= 0xc2 && (c &= 0x1f, 1)) &&
+                // last trail byte
+                (t = *p - 0x80) <= 0x3f) {
+            c = (c << 6) | t;
+            ++length;
+            ++p;
+            return {c, length, true, p0, p};
+        }
+        return {sub(), length, false, p0, p};
+    }
+
+    static inline CodeUnits<CP32, UnitIter> singlePassReadAndInc(
+            UnitIter &p, const UnitIter &limit) {
+        // Very similar to U8_NEXT_OR_FFFD().
+        CP32 c = uint8_t(*p);
+        ++p;
+        if (U8_IS_SINGLE(c)) {
+            return {c, 1, true};
+        }
+        uint8_t length = 1;
+        uint8_t t = 0;
+        if (p != limit &&
+                // fetch/validate/assemble all but last trail byte
+                (c >= 0xe0 ?
+                    (c < 0xf0 ?  // U+0800..U+FFFF except surrogates
+                        U8_LEAD3_T1_BITS[c &= 0xf] & (1 << ((t = *p) >> 5)) &&
+                        (t &= 0x3f, 1)
+                    :  // U+10000..U+10FFFF
+                        (c -= 0xf0) <= 4 &&
+                        U8_LEAD4_T1_BITS[(t = *p) >> 4] & (1 << c) &&
+                        (c = (c << 6) | (t & 0x3f), ++length, ++p != limit) &&
+                        (t = *p - 0x80) <= 0x3f) &&
+                    // valid second-to-last trail byte
+                    (c = (c << 6) | t, ++length, ++p != limit)
+                :  // U+0080..U+07FF
+                    c >= 0xc2 && (c &= 0x1f, 1)) &&
+                // last trail byte
+                (t = *p - 0x80) <= 0x3f) {
+            c = (c << 6) | t;
+            ++length;
+            ++p;
+            return {c, length, true};
+        }
+        return {sub(), length, false};
+    }
+
+    static inline CodeUnits<CP32, UnitIter> decAndRead(UnitIter start, UnitIter &p) {
+        // Very similar to U8_PREV_OR_FFFD().
+        UnitIter p0 = p;
+        CP32 c = uint8_t(*--p);
+        if (U8_IS_SINGLE(c)) {
+            return {c, 1, true, p, p0};
+        }
+        if (U8_IS_TRAIL(c) && p != start) {
+            UnitIter p1 = p;
+            uint8_t b1 = *--p1;
+            if (U8_IS_LEAD(b1)) {
+                if (b1 < 0xe0) {
+                    p = p1;
+                    c = ((b1 - 0xc0) << 6) | (c & 0x3f);
+                    return {c, 2, true, p, p0};
+                } else if (b1 < 0xf0 ?
```

Comment:
As the previous block ends with `return`, it doesn't really make any sense to have an `else`-clause here, this could just be a plain separate `if`-statement.

Reply:
Possible, but it doesn't really matter, and I think the explicit else helps understand the code.

This here is ported from https://github.com/unicode-org/icu/blob/release-77-1/icu4c/source/common/utf_impl.cpp#L243